rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,239 @@
+
+## Detailed Design or Proposal
+
+### A `useMotionPresence` hook based solution
+
+To determine the motion state of a component, a hook could be created as part of the @fluentui/react-utilities package. A preliminary implementation of this hook can be found here: [useMotionPresence](https://github.com/marcosmoura/fluentui/blob/feat/drawer-motion/packages/react-components/react-utilities/src/hooks/useMotionPresence.ts).
+
+#### What is it?
+
+A tracker hook, that monitors the state of animations and transitions for a particular element. A hook to track the state of animations/transitions. This hook does not directly create animations but instead synchronizes with CSS properties to determine the rendering status, visibility, entering, leaving, and ongoing animation of a component. If any CSS changes or properties are overridden, this hook will automatically adjust and stay synchronized.
+
+#### Usage
+
+State:
+
+```ts
+import * as React from 'react';
+import { getNativeElementProps, useMotionPresence, useMergedRefs } from '@fluentui/react-utilities';
+
+import type { SampleProps, SampleState } from './Sample.types';
+
+export const useSample_unstable = ({ open = false }: SampleProps, ref: React.Ref<HTMLElement>): SampleState => {
+  const { ref: componentRef, shouldRender, visible, entering, exiting, animating } = useMotionPresence(open);
+
+  return {
+    components: {
+      root: 'div',
+    },
+
+    root: getNativeElementProps('div', {
+      ref: useMergedRefs(ref, componentRef),
+    }),
+
+    shouldRender,
+    visible,
+    entering,
+    exiting,
+    animating,
+  };
+};
+```
+
+Renderization:
+
+```tsx
+import * as React from 'react';
+import { getSlots } from '@fluentui/react-utilities';
+import type { SampleState, SampleSlots } from './Sample.types';
+
+/**
+ * Render the final JSX of Sample
+ */
+export const renderSample_unstable = (state: SampleState) => {
+  const { slots, slotProps } = getSlots<SampleSlots>(state);
+
+  if (!state.shouldRender) {
+    return null;
+  }
+
+  return <slots.root {...slotProps.root} />;
+};
+```
+
+Styles:
+
+```ts
+import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
+import type { SampleSlots, SampleState } from './Sample.types';
+import type { SlotClassNames } from '@fluentui/react-utilities';
+import { tokens } from '@fluentui/react-theme';
+
+export const SampleClassNames: SlotClassNames<SampleSlots> = {
+  root: 'fui-Sample',
+};
+
+/**
+ * Styles for the root slot
+ */
+const useStyles = makeStyles({
+  root: {
+    opacity: 0,
+    transitionTimingFunction: 'ease-out',
+    transitionProperty: 'opacity',
+    willChange: 'opacity',
+  },
+
+  entering: {
+    transitionDuration: '200ms',
+  },
+
+  exiting: {
+    transitionDuration: '250ms',
+  },
+
```

Comment:
As we discussed and repro-ed offline: I was having problems using multiple durations, when the `exiting` state is applied with a duration that's very different to entering (i.e. 200ms vs 20s) the hook still uses the initial transitionDuration.

Reply:
This was fixed in the implementation branch as well.